#166272330 ch backend setup swagger

GerrardE opened the pull request:
#### What does this PR do?
Set up swagger api docs

#### Description of Task to be completed?
- Define swagger configuration file
- create base swagger doc as yaml file

#### How should this be manually tested?
1. clone this repo
2. run "npm install"
3. run "npm start"
4. Navigate to "localhost:port/api-docs"

#### Any background context you want to provide?
N/A

#### What are the relevant pivotal tracker stories?
[#166272330](https://www.pivotaltracker.com/story/show/166272330)

#### Screenshots (if appropriate)
![image](https://user-images.githubusercontent.com/22410701/58766054-12d18e00-8572-11e9-9d1f-33a16ad46655.png)


#### Questions:
N/A

GerrardE committed:
chore-[166272330]: Setup Swagger API docs

- Install swagger-ui-express and swagger-jsdoc
- define swaggeryaml

[Finishes #166272330]

GerrardE committed:
chore-[166272330]: Setup Swagger API docs

GerrardE committed:
chore-[166272330]: Setup Swagger API docs

GerrardE committed:
chore-[166272330]: Setup Swagger API docs

- Install swagger-ui-express and swagger-jsdoc
- Set up base config file and initial swagger definitions

[Finishes #166272330]

GerrardE committed:
mend

GerrardE committed:
mend

GerrardE committed:
chore-[166272330]: Setup Swagger API docs

- Define swagger configuration file
- create base swagger doc as yaml file

[Finishes #166272330]

fidelisojeah commented on server/config/swagger.js:
```diff
@@ -0,0 +1,23 @@
+    description: 'A Social platform for the creative at heart',
+  },
+  host: 'https://kingsmen-ah-backend-staging.herokuapp.com/',
+  basePath: 'api/api-docs'
```
change to just `/api-docs` if possible

GerrardE replied:
@fidelisojeah done. Thank you for the feedback

GerrardE committed:
mend

fidelisojeah commented on server/routes/api/index.js:
```diff
@@ -1,9 +1,13 @@
 apiRoutes.use('/', userRoutes);
 
+// swagger-ui-express for API endpoint documentation
+apiRoutes.use('/api-docs', swaggerUi.serve, swaggerUi.setup(swaggerSpec));
```
I believe this should be in the `app.js` file then?

GerrardE replied:
@fidelisojeah fixed. Thanks

GerrardE committed:
mend

GerrardE committed:
mend

fidelisojeah commented on swagger.yaml:
```diff
@@ -0,0 +1,98 @@
+        type: "string"
+externalDocs:
+  description: "Find out more about Author's Haven"
+  url: "https://github.com/andela/kingsmen-ah-backend"
```
new line

GerrardE replied:
@fidelisojeah Please review, changes implemented

GerrardE committed:
mend

fidelisojeah commented on swagger.yaml:
```diff
@@ -0,0 +1,99 @@
+      produces:
+      - "application/json"
+      parameters:
+      - in: "body"
```
Can the input parameters be more like:

```yaml
      - name: "email"
        in: "query"
        description: "The email for registration"
        required: true
        type: "string"
        format: "email"
        minLength: 2
        maxLength: 255
      - name: "password"
        in: "query"
        description: "The password for registration"
        required: true
        type: "string",
        format: "password"
        minLength: 2
        maxLength: 255

...
```

This is just a sample. In reality, the swagger parameter definitions should match the validation definitions in https://github.com/andela/kingsmen-ah-backend/blob/develop/server/validations/auth.js

GerrardE replied:
@EMacco @emmygozi @despeauxz @adex001 kindly review my work. Thanks

fidelisojeah replied:
sorry for this, the `"in"` should be `"body"` not `query`

GerrardE committed:
mend

GerrardE committed:
mend

adex001 commented:
Looks good to me

EMacco commented:
I think this is ok.

fidelisojeah commented on swagger.yaml:
```diff
@@ -0,0 +1,154 @@
+      produces:
+      - "application/json"
+      parameters:
+      - name: "username"
```
sorry for this, the `in` should be `"body"` not `"query"`

GerrardE replied:
Alright @fidelisojeah. Thank you for the feedback. I have implemented it.

GerrardE committed:
mend

GerrardE committed:
mend

GerrardE committed:
chore-[166272330]: Setup Swagger API docs

- Pull from develop
- merge updates

[Finishes #166272330]

emmygozi commented:
Great job. Looks good to me so far. The API docs seem well formatted and return the expected response.

You could try adding this `swagger-ui-express` to package.json

GerrardE committed:
chore-[166272330]: Setup Swagger API docs

- ReInstall swaggerjsdoc and swaggeruiexpress
- Ensure tests are passing

[Finishes #166272330]

fidelisojeah commented on swagger.yaml:
```diff
@@ -0,0 +1,124 @@
+      - "application/json"
+      - "application/xml"
+      parameters:
+      - in: "body"
```
please revert to what you had before and use `"formData"` instead of query

GerrardE replied:
Thank you @fidelisojeah. Changes noted

fidelisojeah commented on server/config/swagger.js:
```diff
@@ -0,0 +1,23 @@
+    version: '1.0.0',
+    description: 'A Social platform for the creative at heart',
+  },
+  host: 'kingsmen-ah-backend-staging.herokuapp.com',
```
I don't want this hardcoded.

use: `host: process.env.HOST_NAME || 'localhost:3000'`

then modify the `app.json` file to look like this:

```json
{
  "name": "kingsmen-ah-backend",
  "description": "A community of like minded authors that fosters inspiration and innovation by leveraging the modern web.",
  "scripts": {
    "postdeploy": "heroku config:set HOST_NAME=${HEROKU_APP_NAME}.herokuapp.com --app ${HEROKU_APP_NAME} && npm run refreshdb"
  },
  "env": {
    "HEROKU_APP_NAME": {
      "required": true
    },
    "HEROKU_PARENT_APP_NAME": {
      "required": true
    }
  },
  "formation": {},
  "addons": [],
  "buildpacks": [
    {
      "url": "heroku/nodejs"
    }
  ],
  "stack": "heroku-18"
}
```

GerrardE replied:
Thankyou @fidelisojeah. 

Here: "heroku config:set HOST_NAME=${HEROKU_APP_NAME}.herokuapp.com --app ${HEROKU_APP_NAME} && npm run refreshdb"
If i am right, this means heroku will check env config for this ${HEROKU_APP_NAME} value

GerrardE committed:
chore-[166272330]: Setup Swagger API docs

- Refactor swaggeryaml and swagger config file
- modify appjson file

[Finishes #166272330]

fidelisojeah commented on app.json:
```diff
@@ -1,12 +1,23 @@
+    }
+  ],
   "stack": "heroku-18"
 }
```
?

GerrardE replied:
Fixed

GerrardE committed:
chore-[166272330]: Setup Swagger API docs

- Refactor swaggeryaml and swagger config file
- modify appjson file

[Finishes #166272330]

GerrardE committed:
chore-[166272330]: Setup Swagger API docs

- Modify appjson file to include space at the end of the file

[Finishes #166272330]

GerrardE committed:
chore-[166272330]: Setup Swagger API docs

- Modify swaggerjs as flagged by houndCI

[Finishes #166272330]

GerrardE commented on server/config/swagger.js:
```diff
@@ -0,0 +1,27 @@
+import swaggerJSDoc from 'swagger-jsdoc';
```
Fixed

EMacco reviewed (dismissed):
Works fine. Well done

fidelisojeah commented on swagger.yaml:
```diff
@@ -0,0 +1,154 @@
+  license:
+    name: "MIT"
+    url: "https://www.apache.org/licenses/LICENSE-2.0.html"
+host: "kingsmen-ah-backend-staging.herokuapp.com"
```
I remember commenting that you should remove this

GerrardE replied:
Hello, @fidelisojeah I removed it. Remember you flagged it on the swagger.json configuration file. this file is the swagger.yaml file and it returns structural error if it is done like so

<img width="719" alt="Screenshot 2019-06-06 at 11 55 14 AM" src="https://user-images.githubusercontent.com/22410701/59027888-19058a00-8852-11e9-9300-b6b1a8516ce0.png">
 
Please advise

fidelisojeah replied:
You understand what `remove` means yes? and you can see the file I am commenting on?

fidelisojeah commented on server/config/swagger.js:
```diff
@@ -0,0 +1,27 @@
+    version: '1.0.0',
+    description: 'A Social platform for the creative at heart',
+  },
+  host: `${process.env.HOST_NAME}` || 'localhost:3000',
```
placing this in a string like this This causes a problem if `process.env.HOST_NAME` does not exist we have a string `'undefined'`  which is a valid string.

I fail to understand why you have this like this despite my telling you otherwise

fidelisojeah commented on package.json:
```diff
@@ -33,18 +33,20 @@
     "sequelize-cli": "^5.4.0",
-    "sqlite3": "^4.0.8"
+    "sqlite3": "^4.0.8",
+    "swagger-jsdoc": "^3.2.9",
```
Are you using `swagger-jsdoc` ?

GerrardE replied:
Yes i am

GerrardE committed:
chore-[166272330]: Setup Swagger API docs

- Modify swagger js/yaml files

[Finishes #166272330]

GerrardE committed:
Merge branch 'develop' into 166272330-ch-backend-setup-swagger

GerrardE committed:
chore-[166272330]: Setup Swagger API docs

- merge updates from develop

[Finishes #166272330]

GerrardE committed:
chore-[166272330]: Setup Swagger API docs

- update appjson to include buildpack for heroku-cli

[Finishes #166272330]

GerrardE committed:
chore-[166272330]: Setup Swagger API docs

- update appjson to include buildpack for heroku-cli

[Finishes #166272330]

GerrardE committed:
chore-[166272330]: Setup Swagger API docs

- update appjson and swaggerjs

[Finishes #166272330]

GerrardE committed:
chore-[166272330]: Setup Swagger API docs

- update appjson and swaggerjs

[Finishes #166272330]

despeauxz reviewed (approved):
It works fine now. good job
